### ISSUE TITLE
Job submission rate throttling needed for AWS Batch

### ISSUE THREAD

**Issue by sminot** [nextflow-io/nextflow#1371]
### Expected behavior and actual behavior

In the past, I have not had any problems with the job submission rate while using the AWS Batch executor. Since switching to 19.10.0, I have started to frequently get 503 errors from AWS Batch (`Please reduce your request rate. (Service: Amazon S3; Status Code: 503; Error Code: SlowDown`). 

I know that this has been addressed previously, but I wanted to open this issue because I believe that the earlier solutions may no longer be working (either due to a change on the AWS side or locally).

### Steps to reproduce the problem

Using 19.10.0 with a job submission rate of 1 / sec, I consistently get `Please reduce your request rate. (Service: Amazon S3; Status Code: 503; Error Code: SlowDown` errors when submitting >1 job. 

### Environment 

* Nextflow version: 19.10.0
* Java version: java version "1.7.0_151", OpenJDK Runtime Environment (IcedTea 2.6.11) (7u151-2.6.11-2ubuntu0.14.04.1), OpenJDK 64-Bit Server VM (build 24.151-b01, mixed mode)
* Operating system: Ubuntu

**Comment by pditommaso** [comment]
Weird. Are you able to replicate the problem? or any clue which version is working fine?

**Comment by sminot** [comment]
It's weirder than that even -- I'm getting the error now even when I don't submit any jobs. 

Are all of the API calls being monitored for rate throttling, not just the job submissions?

I see the error on 19.09.0-edge as well as 19.10.0. I haven't tried anything lower because I really want to use Tower.

It's also possible that the error is because something changed on the AWS side, I suppose.

**Comment by pditommaso** [comment]
Can you upload a `.nextflow.log` reporting the error?

**Comment by sminot** [comment]
I'm managed to isolate the error and figure out that it only happens when I'm running two rather large workflows concurrently on the same machine. When I run either one individually, it works fine, but when I run them at the same time they both fail.

The log file is rather large -- would it be helpful if I passed along the last N lines?

**Comment by tbugfinder** [comment]
Is the submit host also on AWS, if yes which type is it?
Did you check memory use?

**Comment by pditommaso** [comment]
> I'm managed to isolate the error and figure out that it only happens when I'm running two rather large workflows concurrently on the same machine. 

I guess AWS cap the requests/rate by the source IP, you should launch different NF runs on different machines. Don't think there's much we can do on NF side, this is something that will be addressed by Tower. 

Anyhow still can be useful to have a look to the log file, the relevant part is where there's the error stack trace (I guess to the bottom).

**Comment by pditommaso** [comment]
Closing this since this is mainly due to AWS limits.

**Comment by JonathanCSmith** [comment]
I am also observing this behaviour - could we not have a catch on this type of failure with an attempt to resub after a period of time? For embarrassingly parallel workloads at scale this is a real issue...

**Comment by sminot** [comment]
I wanted to reopen this issue, as I am finding that it happens reproducibly with a workflow that is processing ~4000 samples in parallel.

```
Please reduce your request rate. (Service: Amazon S3; Status Code: 503; Error Code: SlowDown; Request ID:---; S3 Extended Request ID: ---)
```

Here's a shorter version of the end of the log file. The full file is 2Mb, not sure if I should post it here.

```
Feb-18 23:20:33.934 [Actor Thread 119] INFO  nextflow.processor.TaskProcessor - [b1/3d5eb9] Cached process > combineReads (Join FASTQ files per-specimen)
Feb-18 23:20:34.177 [main] DEBUG nextflow.trace.TraceFileObserver - Flow completing -- flushing trace file
Feb-18 23:20:34.179 [main] DEBUG nextflow.trace.ReportObserver - Flow completing -- rendering html report
Feb-18 23:20:35.578 [Actor Thread 132] INFO  nextflow.processor.TaskProcessor - [bd/6a1991] Cached process > assembly_wf:assembly (De novo metagenomic assembly)
Feb-18 23:20:35.754 [main] DEBUG nextflow.CacheDB - Closing CacheDB done
Feb-18 23:20:35.756 [main] DEBUG nextflow.Session - AWS S3 uploader shutdown
Feb-18 23:20:35.776 [main] DEBUG nextflow.script.ScriptRunner - > Execution complete -- Goodbye
```

Can we reopen and address this, @pditommaso ?

**Comment by JonathanCSmith** [comment]
I managed to get over most of these issues with some combination of the following attributes in my nextflow.config. Additionally, care must be taken to ensure that the amount of cpu threads is not exhausted on the uploading machine (key property being uploadMaxThreads and maxConnections). This requires a nextflow version > 19.07.

```
profiles {
	    aws {
	        params.debug = true
	        process.executor = 'awsbatch'
	        process.queue = ''

	        executor {
	            submitter {
	                rateLimit = '20/s'
	                autoThrottle = true
	                poolSize = 8
	                queueSize = 5000
	                maxPoolSize = 8
	                maxRetries = 50000
	                keepAlive = '1 min'
	                errorBurstDelay = '1 sec'
	                backOffMinRate = 0
	                backOffFactor = 2
	                rampUpInterval = 100
	                rampUpFactor = 1.2
	            }
	        }

	        aws {
	            accessKey = ''
	            secretKey = ''
	            region = ''

	            client {
	                protocol = 'https'
	                uploadStorageClass = 'INTELLIGENT_TIERING'
	                maxConnections = 32 // This seems to be closely tied to uploadMaxThreads - vast differences caused issues for me
	                maxErrorRetry = 100
	                uploadMaxThreads = 32 // Vary this according to your uploader computer
	                uploadChunkSize = '100MB' // I have found larger chunk sizes to be more stable
	                uploadMaxAttempts = 10
	                uploadRetrySleep = '10 sec'
	            }

	            batch {
	                cliPath = '/home/ec2-user/miniconda/bin/aws'
	                maxTransferAttempts = 100
	                delayBetweenAttempts = 1000
	                maxParallelTransfers = 8
	            }
	        }
	    }
	}
}
```

**Comment by sminot** [comment]
@JonathanCSmith in your experience did this help with errors coming from AWS Batch or AWS S3?

The errors I'm getting are from S3, and I wonder if limiting the submission rate to Batch is going to help. 

My error comes in a case when a large number of jobs have been cached, and the job gets killed while it's still checking all of the cached jobs. Is that the same as what you have encountered?

**Comment by pditommaso** [comment]
Something I didn't realise originally is that this error is reported by S3 not Batch. This makes more sense because there's an auto-throttling mechanism for Batch. 

The use for S3 is intrinsically limited for such workload. We have seen also problem saturating all node bandwidth while downloading/uploading data.

The plan is to support FSx for Lustre for large parallel workloads.

**Comment by sminot** [comment]
I can confirm that my immediate problem has been solved by adding:

```
aws {
    client {
        maxConnections = 4
    }
}
```

**Comment by jwarwick-delfi** [comment]
We hit this exact issue trying to run lots of nextflow pipelines at once on Batch. In talking with AWS support, it turns out that we were getting the 503 "Slow Down" error from S3 _not because of the rate of our requests_, which was far below the limit, but because nextflow was trying to write the same file at the same time. From our support chat:

> The reason why you're seeing a 503 despite being far from the 3500 PUTs/sec limit is because there are multiple PUTs happening on this key at the same time.
For example, if you do two PUTs on the same object key at the same time, S3 cannot ensure Strong Consistency so it throws a 503 for the second request until it can ensure Strong Consistency for the first request.
To resolve this, the application shouldn't make multiple PUT requests on the same object key within a few milliseconds of each other

We also learned that the file collision was happening in the nextflow work directory. Our workflow has some processes that run in parallel / are launched at the same time, which might explain the concurrent access to the same files, I guess... @pditommaso - does nextflow use the same ID and work directory when the same process is executed multiple times?

We also learned that the conflict was occurring on a directory, not a specific file: `s3://(bucket)/nextflow-work/stage/98/a2e38c9aabc0b71519cdc2aeb26017/`. I am not sure how that could be - wild stab in the dark here, is nextflow using s3fs which might be using a put request to ensure the directory exists?

Setting maxForks to 1 would work, but would also limit us to launching one at a time.
